File: src/nonogram/web/static/metadata.js
/**
 * Client-side image metadata calculation and dimension suggestions (CARD-034).
 *
 * This module provides real-time calculation of image dimensions and puzzle
 * dimension suggestions when a user selects an image file. The algorithm
 * matches the server-side implementation exactly (AC-137).
 *
 * AC-135: Metadata and suggestions appear instantly when file selected
 * AC-136: Clicking suggestion populates size field, file input retains selection
 * AC-137: Client-side calculations match server-side exactly
 * AC-138: Graceful fallback if File API unavailable
 */

(function() {
  "use strict";

  /**
   * Calculate the greatest common divisor of two numbers.
   * Used to simplify aspect ratios.
   */
  function gcd(a, b) {
    return b === 0 ? a : gcd(b, a % b);
  }

  /**
   * Simplify a ratio to its lowest terms.
   * Returns [simplified_width, simplified_height].
   */
  function simplifyRatio(width, height) {
    const divisor = gcd(width, height);
    return [width / divisor, height / divisor];
  }

  /**
   * Format an aspect ratio for display.
   * Returns a string like "4:3 (1.33)".
   */
  function formatAspectRatio(width, height, decimal) {
    return `${width}:${height} (${decimal})`;
  }

  /**
   * Extract metadata from an image file.
   * Returns a Promise that resolves with {width, height, aspectRatio, decimal}
   * or rejects if the image cannot be read.
   *
   * AC-135: Instant metadata calculation when file is selected
   */
  function extractImageMetadata(file) {
    return new Promise((resolve, reject) => {
      // Create a FileReader to read the image
      const reader = new FileReader();

      reader.onload = function(e) {
        // Create an Image object to get dimensions
        const img = new Image();

        img.onload = function() {
          const [simplifiedW, simplifiedH] = simplifyRatio(img.width, img.height);
          const decimal = Math.round((img.width / img.height) * 100) / 100;

          resolve({
            width: img.width,
            height: img.height,
            aspectRatio: {
              width: simplifiedW,
              height: simplifiedH,
              decimal: decimal
            }
          });
        };

        img.onerror = function() {
          reject(new Error("Could not read image dimensions"));
        };

        // Set the image source to trigger loading
        img.src = e.target.result;
      };

      reader.onerror = function() {
        reject(new Error("Could not read file"));
      };

      // Read the file as a data URL
      reader.readAsDataURL(file);
    });
  }

  /**
   * Generate 2-3 suggested puzzle dimensions based on image aspect ratio.
   * Matches the server-side algorithm in metadata.py exactly (AC-137).
   *
   * Returns a list of [width, height] pairs, ordered by how closely they
   * match the aspect ratio. All dimensions are within [minSize, maxSize].
   */
  function suggestDimensions(metadata, minSize = 10, maxSize = 30) {
    const aspectW = metadata.aspectRatio.width;
    const aspectH = metadata.aspectRatio.height;
    const targetRatio = aspectW / aspectH;

    const suggestions = [];

    // Try all combinations within the constraint
    for (let w = minSize; w <= maxSize; w++) {
      for (let h = minSize; h <= maxSize; h++) {
        const gridRatio = w / h;
        // Calculate how close this dimension is to the target aspect ratio
        const ratioError = Math.abs(gridRatio - targetRatio) / targetRatio;
        suggestions.push([ratioError, [w, h]]);
      }
    }

    // Sort by ratio error (closest first)
    suggestions.sort((a, b) => a[0] - b[0]);

    // Return top 2-3 suggestions
    const resultCount = Math.min(3, suggestions.length);
    return suggestions.slice(0, resultCount).map(item => item[1]);
  }

  /**
   * Update the form with metadata and suggestions.
   * AC-135: Display metadata and suggestions in real time
   * AC-136: Clicking suggestion buttons populates the size field
   * AC-158, AC-159: Display image preview with dimensions
   */
  function updateFormWithMetadata(metadata, suggestions) {
    // Format the aspect ratio for display
    const aspectRatioStr = formatAspectRatio(
      metadata.aspectRatio.width,
      metadata.aspectRatio.height,
      metadata.aspectRatio.decimal
    );

    // Display image preview (AC-158, AC-159, AC-160)
    displayImagePreview(metadata);

    // Create metadata section HTML
    const metadataHtml = `
      <div class="metadata">
        <p><strong>Image aspect ratio:</strong> ${escapeHtml(aspectRatioStr)}</p>
      </div>
    `;

    // Create suggestions section HTML
    let suggestionsHtml = "";
    if (suggestions && suggestions.length > 0) {
      const buttons = suggestions
        .map(([w, h]) => {
          const sizeStr = `${w}x${h}`;
          return `
            <button type="button" class="suggestion-button" data-size="${escapeHtml(sizeStr)}">${escapeHtml(sizeStr)}</button>
          `;
        })
        .join("");

      suggestionsHtml = `
        <div class="suggestions">
          <p><small><strong>Suggested dimensions (click to set):</strong></small></p>
          ${buttons}
        </div>
      `;
    }

    // Find the metadata-suggestions-area container
    const container = document.getElementById("metadata-suggestions-area");
    if (container) {
      container.innerHTML = metadataHtml + suggestionsHtml;

      // Attach click handlers to suggestion buttons
      const buttons = container.querySelectorAll(".suggestion-button");
      buttons.forEach(function(button) {
        button.addEventListener("click", function(e) {
          e.preventDefault();
          const sizeStr = this.getAttribute("data-size");
          const sizeInput = document.querySelector('input[name="size"]');
          if (sizeInput) {
            sizeInput.value = sizeStr;
          }
          return false;
        });
      });
    }
  }

  /**
   * Display the image preview thumbnail (AC-158, AC-159, AC-160).
   * Shows a thumbnail (max 150×150 px) with original dimensions label.
   */
  function displayImagePreview(metadata) {
    const previewContainer = document.getElementById("image-preview-container");
    const previewImg = document.getElementById("image-preview");
    const dimensionsDiv = document.getElementById("image-dimensions");

    if (!previewContainer || !previewImg || !dimensionsDiv) {
      return;
    }

    // Use FileReader to get the data URL for preview
    const fileInput = document.querySelector('input[type="file"][name="image"]');
    if (!fileInput || fileInput.files.length === 0) {
      return;
    }

    const file = fileInput.files[0];
    const reader = new FileReader();

    reader.onload = function(e) {
      previewImg.src = e.target.result;
      dimensionsDiv.textContent = `${metadata.width} × ${metadata.height}`;
      previewContainer.classList.add("visible");
    };

    reader.onerror = function() {
      previewContainer.classList.remove("visible");
    };

    reader.readAsDataURL(file);
  }

  /**
   * Simple HTML escape function to prevent XSS.
   * AC-138: Graceful handling - only escape user data
   */
  function escapeHtml(text) {
    if (typeof text !== "string") {
      return "";
    }
    const div = document.createElement("div");
    div.textContent = text;
    return div.innerHTML;
  }

  /**
   * Clear metadata and suggestions from the form.
   */
  function clearMetadata() {
    const container = document.getElementById("metadata-suggestions-area");
    if (container) {
      container.innerHTML = "";
    }

    // Clear preview (AC-160: update when new file selected)
    const previewContainer = document.getElementById("image-preview-container");
    if (previewContainer) {
      previewContainer.classList.remove("visible");
    }
  }

  /**
   * Clear the size field (AC-149: clear size when new image selected).
   * Prevents stale size from previous image being used with new image.
   */
  function clearSizeField() {
    const sizeInput = document.querySelector('input[name="size"]');
    if (sizeInput) {
      sizeInput.value = "";
    }
  }

  /**
   * Initialize the file input listener.
   * AC-135: Set up handler for file selection
   * AC-138: Graceful fallback if File API unavailable
   */
  function initializeFileInputListener() {
    try {
      const fileInput = document.querySelector('input[type="file"][name="image"]');

      if (!fileInput) {
        // File input not found - graceful fallback (AC-138)
        console.log("Image file input not found - metadata calculation unavailable");
        return;
      }

      // Check if File API is available (AC-138)
      if (!window.FileReader || !window.Image) {
        console.log("File API not available - metadata calculation unavailable");
        fileInput.addEventListener("change", function() {
          clearMetadata();
        });
        return;
      }

      // AC-135: Handle file selection and calculate metadata
      fileInput.addEventListener("change", function() {
        try {
          if (this.files.length === 0) {
            clearMetadata();
            clearSizeField(); // AC-149: Clear size field when no image selected
            return;
          }

          const file = this.files[0];

          // Verify it's an image file
          if (!file.type.startsWith("image/")) {
            clearMetadata();
            clearSizeField(); // AC-149: Clear size field for invalid image
            return;
          }

          // AC-149: Clear size field when new image selected (prevents stale size)
          clearSizeField();

          // AC-135: Extract metadata asynchronously
          extractImageMetadata(file)
            .then(function(metadata) {
              // AC-135: Calculate suggestions
              const suggestions = suggestDimensions(metadata);

              // AC-135: Update form with results
              updateFormWithMetadata(metadata, suggestions);
            })
            .catch(function(error) {
              // AC-138: Graceful error handling - clear metadata on error
              console.log("Could not extract image metadata:", error.message);
              clearMetadata();
            });
        } catch (err) {
          // AC-138: Graceful error handling
          console.log("Error processing file change:", err.message);
          clearMetadata();
          clearSizeField(); // AC-149: Clear size on error
        }
      });
    } catch (err) {
      // AC-138: Graceful fallback if initialization fails
      console.log("Could not initialize metadata listener:", err.message);
    }
  }

  // Initialize when DOM is ready
  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", initializeFileInputListener);
  } else {
    // DOM already loaded
    initializeFileInputListener();
  }
})();

File: src/nonogram/web/metadata.py
"""Image metadata extraction and puzzle dimension suggestions (CARD-031).

This module extracts metadata from uploaded images and generates suggestions
for puzzle dimensions that fit the image's aspect ratio within the 10..30
constraint (CON-011).
"""

from __future__ import annotations

from dataclasses import dataclass
from math import gcd
from pathlib import Path
from typing import NamedTuple

from nonogram.sourcing import image as sourcing_image

__all__ = [
    "AspectRatio",
    "ImageMetadata",
    "extract_metadata",
    "suggest_dimensions",
]


class AspectRatio(NamedTuple):
    """A simplified aspect ratio representation.

    Attributes:
        width: The width component of the ratio (simplified).
        height: The height component of the ratio (simplified).
        decimal: The decimal representation (width / height).
    """

    width: int
    height: int
    decimal: float


@dataclass(frozen=True, slots=True)
class ImageMetadata:
    """Metadata extracted from an uploaded image.

    Attributes:
        width: The image width in pixels (ink bounding box).
        height: The image height in pixels (ink bounding box).
        aspect_ratio: The simplified aspect ratio.
    """

    width: int
    height: int
    aspect_ratio: AspectRatio


def _simplify_ratio(width: int, height: int) -> tuple[int, int]:
    """Simplify a ratio to its lowest terms.

    Args:
        width: The width component.
        height: The height component.

    Returns:
        (simplified_width, simplified_height)
    """
    divisor = gcd(width, height)
    return width // divisor, height // divisor


def extract_metadata(image_path: Path) -> ImageMetadata:
    """Extract metadata from an uploaded image.

    Args:
        image_path: Path to the uploaded image file.

    Returns:
        ImageMetadata with dimensions and aspect ratio.

    Raises:
        UnreadableImage: If the image cannot be read or decoded.
    """
    width, height = sourcing_image.source_shape(str(image_path))
    simplified_w, simplified_h = _simplify_ratio(width, height)
    aspect_ratio = AspectRatio(
        width=simplified_w,
        height=simplified_h,
        decimal=round(width / height, 2),
    )
    return ImageMetadata(width=width, height=height, aspect_ratio=aspect_ratio)


def suggest_dimensions(
    metadata: ImageMetadata, min_size: int = 10, max_size: int = 30
) -> list[tuple[int, int]]:
    """Generate 2-3 suggested puzzle dimensions based on image aspect ratio.

    The suggestions fit the image's aspect ratio within the min_size..max_size
    constraint. Dimensions are ordered by how closely they match the aspect
    ratio (closest first).

    Args:
        metadata: The extracted image metadata.
        min_size: Minimum grid dimension (default 10, per CON-011).
        max_size: Maximum grid dimension (default 30, per CON-011).

    Returns:
        A list of (width, height) tuples, 2-3 items long, ordered by ratio
        match quality. All dimensions are within [min_size, max_size].
    """
    aspect_w = metadata.aspect_ratio.width
    aspect_h = metadata.aspect_ratio.height
    target_ratio = aspect_w / aspect_h

    suggestions: list[tuple[float, tuple[int, int]]] = []

    # Try all combinations within the constraint
    for w in range(min_size, max_size + 1):
        for h in range(min_size, max_size + 1):
            grid_ratio = w / h
            # Calculate how close this dimension is to the target aspect ratio
            ratio_error = abs(grid_ratio - target_ratio) / target_ratio
            suggestions.append((ratio_error, (w, h)))

    # Sort by ratio error (closest first) and take 2-3 best
    suggestions.sort(key=lambda x: x[0])

    # Return top 2-3 suggestions (prefer 3 if available)
    result_count = min(3, len(suggestions))
    return [dims for _, dims in suggestions[:result_count]]


def format_aspect_ratio(aspect_ratio: AspectRatio) -> str:
    """Format aspect ratio as a string for display.

    Args:
        aspect_ratio: The aspect ratio to format.

    Returns:
        A string like "4:3 (1.33)".
    """
    return f"{aspect_ratio.width}:{aspect_ratio.height} ({aspect_ratio.decimal})"
